### PR TITLE
[RFC #32321] Serve Apple Silicon decode through one exported whole-model MLX region

### PR DESCRIPTION
**Pull request by noob-se7en**
## Motivation

**RFC: #32321 — this PR is the implementation of the whole-region proposal in that RFC; please read it first for the design rationale and the comparison with the provider-based alternative.**

SGLang on Apple Silicon currently has two serving architectures: the standard Torch `ModelRunner` path, and a parallel native-MLX lane (`SGLANG_USE_MLX` → `MlxTpModelWorker` → `MlxModelRunner`) with its own model load, KV cache, and request lifecycle. The RFC argues for one authoritative owner: the Torch `ModelRunner` keeps scheduling, pools, sampling, and logits, and MLX becomes a bounded compute region under it — captured from the loaded Torch model with non-strict `torch.export` and lowered to MLX per ATen op. No model adapters, no second model load, no parallel serving state.

This PR adds that region as an additive, off-by-default provider (`SGLANG_ENABLE_MLX_WHOLE_REGION=1`); anything the region cannot serve falls back to the eager Torch MPS path per batch. Removing the native MLX lane is deliberately a follow-up PR, once behavior and performance parity are agreed. Earlier iterations with full measurement history: noob-se7en#2 (full RFC branch), noob-se7en#3 (this extraction).

## Modifications

Six commits: one per layer, then three serving refinements from review (startup export, bucket ladders, registration):

1. **Capture** — `layers/radix_attention.py` emits the semantic attention op (`sglang::unified_attention_with_output`, carrying `k_cache`/`v_cache`/`req_to_token`/`seq_lens`) when `torch.compiler.is_exporting()` is true; eager execution is unchanged, and `k_cache`/`v_cache` are declared mutated so functionalization cannot elide the pool write. `ServingForwardExportWrapper` resolves the decoder body structurally (single RadixAttention-bearing layer stack, arguments bound by name), so no per-model adapter exists, and narrows logits to `config.vocab_size` exactly as `LogitsProcessor` does.
2. **Lowering** — `hardware_backend/mlx/fx_backend.py` walks the exported graph once; generic ATen ops map to `mx` ops (`mx.fast.rms_norm`/`layer_norm`; `mx.fast.scaled_dot_product_attention` for single-request no-prefix prefill). The attention op lowers to a deferred radix-attention Metal kernel templated on the KV-pool head geometry; pool buffers bind as persistent zero-copy DLPack views of Torch MPS storage. The two hand-written kernels exist only where the paged pool makes `mx.fast` inapplicable: attention reading K/V through `req_to_token`, and the KV-delta commit scatter (MLX kernels cannot mutate Torch-owned buffers; `verify_deferred_kv_commit` probes the real pool through the commit kernel at executor construction and fails loudly on silently dropped writes).
3. **Serving** — `hardware_backend/mlx/region_runner.py` serves decode plus single-request prefill. It is registered as the `mps` entry of `graph_runners` in `model_runner_components/cuda_graph_setup.py` (the decode slot; the same instance is shared into the prefill slot) and is driven by `cuda_graph_config` like the CPU/NPU/XPU runners: both ladders are exported at startup and batches pad onto the smallest covering bucket as CUDA-graph replay does — decode batch sizes (default 1, 2, 4, 8, 12, 16) and prefill token buckets (default 128–2048), set with the standard `--cuda-graph-bs-{decode,prefill}` / `--cuda-graph-max-bs-{decode,prefill}` flags. Pads use the pool's reserved rows (req 0, K/V to padding slot 0) and their logits are dropped. Models outside the kernel contract are rejected at startup with a logged reason (uniform bf16 KV geometry, head_dim multiple of the simdgroup width, `head_dim ** -0.5` scaling, no sliding window, no logit capping, no nontrivial `LogitsProcessor`); a failed export blacklists that batch shape; pool reallocation or weight replacement invalidates cached executors via a storage-identity check. Spec decode, LoRA, encoder batches, logprobs, and multi-request prefill are guarded to the eager Torch path.

Not in this PR: native-lane removal (follow-up), eager-path Metal ops, LoRA-on-MPS fixes, the region validation harness (lives on the RFC branch).

## How to run

```bash
SGLANG_ENABLE_MLX_WHOLE_REGION=1 python -m sglang.launch_server \
  --model-path Qwen/Qwen3-0.6B --host 127.0.0.1 --port 30000

# A/B against the eager Torch MPS path (flag off vs on)
SGLANG_ENABLE_MLX_WHOLE_REGION=1 python -m sglang.bench_one_batch \
  --model-path Qwen/Qwen3-0.6B --batch-size 8 --input-len 512 --output-len 65

# Smaller export ladders: the standard CUDA-graph flags (MPS-sized defaults otherwise)
SGLANG_ENABLE_MLX_WHOLE_REGION=1 python -m sglang.launch_server \
  --model-path Qwen/Qwen3-0.6B --cuda-graph-max-bs-decode 8 --cuda-graph-max-bs-prefill 512
```

Both ladders are exported at startup, like CUDA-graph capture; nothing exports in the request path. They are the standard CUDA-graph flags with MPS-sized defaults — `--cuda-graph-bs-decode` / `--cuda-graph-max-bs-decode` (default 1, 2, 4, 8, 12, 16) and `--cuda-graph-bs-prefill` / `--cuda-graph-max-bs-prefill` (token counts; default 128, 192, 256, 384, 512, 768, 1024, 1536, 2048) — and `--cuda-graph-backend-{decode,prefill}=disabled` turns a phase off. An export costs ≈2 s per shape, so the default 15 shapes take ≈35–45 s at boot on Qwen3-0.6B; the caps bound it. Startup export time is reported in the graph-capture log line.

## How the region runner plugs in, and the shape lifecycle

`MlxRegionRunner` is not a capture/replay graph runner and deliberately does not subclass `DecodeCudaGraphRunner` (the NPU/XPU pattern): those runners swap a device graph object, capture context and `replay()` (`torch.npu.NPUGraph`, `torch.xpu.XPUGraph`) into the shared fixed-buffer/replay machinery, and `torch.mps` exposes none of the three. The region is the compile-based sibling — `torch.export` + MLX lowering + `mx.compile`, a function called with each step's tensors — the same family as the `torch.compile` `CPUGraphRunner`, and it fulfills the slot contract at `BaseRunner` (`can_run_graph` / `load_batch` / `execute`), the level `ModelRunner`'s dispatch depends on. It is registered as `"mps"` in the device→runner map inside `capture_decode_graph`, like the CPU/NPU/XPU runners, and `cuda_graph_config` is live on MPS: with `SGLANG_ENABLE_MLX_WHOLE_REGION=1`, resolution keeps both phases at `Backend.FULL` — the region is one whole-forward executor per bucket for both phases, which is exactly what `full` means — instead of auto-disabling graphs for the `torch_native` attention backend, and fills MPS-sized ladders (`hardware_backend/mlx/region_config.py`, pure Python so `server_args` can import it: same flags, platform-sized defaults). Prefill has no device registry, so `capture_cuda_graphs` shares the decode-registered instance into the prefill slot; one instance dispatches on forward mode. Draft workers construct one too, and it declines them.

**Startup export (CUDA-graph parity).** Like the CUDA runners, construction exports every bucket of both ladders before serving — no level knob, no lazy path for serving shapes (default 6 decode + 9 prefill executors, ≈35–45 s at boot on Qwen3-0.6B). Synthetic batches point at the pool's reserved rows (req 0 / slot 0, the same rows CUDA-graph capture aims dummy batches at), one warm execution per shape triggers the MLX graph compile, `mx.clear_cache()` after each shape bounds the buffer-cache footprint, and the elapsed time is reported through `GraphCapture.capture_time`.

Per batch:

1. Serving dispatch calls `can_run_graph(forward_batch)`, exactly as for CUDA graphs. Spec decode, LoRA, encoder batches, logprobs, multi-request prefill, and out-of-contract models answer false and serve eagerly.
2. The runner computes an executor key: `("decode", bucket)` — the batch size rounds up to the smallest configured decode bucket and the batch pads to it, as CUDA-graph replay does (pad rows are the reserved dummy request: req 0, sequence length 1, K/V to padding slot 0; their logits are dropped) — or `("extend", bucket)` for prefill, where the prompt length rounds up to the configured token ladder and the batch pads to it (pads are causal-safe, their K/V commits to slot 0, and serving slices the last real token's logits). Above the largest bucket of either ladder the batch serves eagerly.
3. **Every batch**: cache hit on the bucket's executor, exported at startup — nothing exports in the request path. All step-to-step variability — token ids, positions, cache slots, sequence lengths — enters through tensor arguments; a baked-constant independence check verifies the traced graph does not depend on the Python-scalar batch fields, which is what makes reuse sound, across steps and across the padded batches that share a bucket.
4. A shape whose export fails is blacklisted and serves eagerly from then on; KV-pool reallocation or weight replacement invalidates all cached executors through a storage-identity check, after which they re-export on their next batch.

Verified end to end on Qwen3-0.6B: all 15 default executors exist before the first request, the decode slot is an `MlxRegionRunner` built through `graph_runners`, `cuda_graph_config` resolves to `full`/`full`, a bs=3 decode rides the bs=4 executor with exact token match against eager over consecutive steps (the second attending over K/V the first's padded commit wrote), `--cuda-graph-bs-decode 6 2` routes it onto 6, and batches above the cap serve eagerly.

## Accuracy Tests

- End-to-end through the real `ModelRunner`/`ForwardBatch` path, greedy, flag on vs off: **exact token match** on four architectures with no model-specific code — Qwen3-0.6B (qwen3), gpt2 (padded vocab 50257→50304), opt-125m (padded vocab, `pp_proxy` body), phi-1_5 (swapped body signature). Gemma2 is correctly rejected at startup and serves eagerly. Qwen2.5-Coder-1.5B and SmolLM2-360M additionally verified on the RFC branch (17-step lockstep parity).
- `test_serving_boundary_parity` uses `LogitsProcessor` itself as the oracle on a deliberately padded-vocabulary fixture, so wrapper-vs-serving divergence cannot cancel out of the comparison.
- Unit suites (77 tests + 13 subtests, green on Apple M3 Pro): topology resolver and body binding, pre-launch kernel contract, region gating and executor lifecycle, startup export of both ladders, decode bucket routing / pad-row contents / logit slicing, ladders read from `cuda_graph_config` with the request-pool clamp, the MPS graph-config policy (MPS-sized defaults, explicit lists and caps, per-phase disable, unsupported backends rejected), lowering registry, deferred attention kernel vs eager reference, KV-commit kernel vs the real pool layout, runtime-version gate.

## Speed Tests and Profiling

Measured on the RFC branch (noob-se7en#2; same region code; Apple M3 Pro 36 GB, Qwen3-0.6B bf16, 512 in / 65 out, real scheduler, medians of warm runs): **3.2–3.6× end-to-end output tok/s** over the eager Torch MPS path from batch 1 to 16 (18.1→58.7, 49.0→167.8, 57.4→204.9 tok/s at batch 1/8/16), TTFT 318→233 ms. At the region boundary the exported decode forward runs at ~93 % of a native `mlx_lm` forward of the same shape; prefill ~89 % via `mx.fast` causal attention. Export cost is ≈2 s per shape, paid once at startup for the whole ladder; no request pays it. Reproduce the end-to-end A/B with:

```bash
SGLANG_ENABLE_MLX_WHOLE_REGION=1 python -m sglang.bench_one_batch \
  --model-path Qwen/Qwen3-0.6B --batch-size 8 --input-len 512 --output-len 65
```

## Known gaps

- Prefill serving is single-request only (packed multi-request prefill stays on Torch); prompt logprobs are guarded to Torch.
- LoRA and speculative decoding are guarded out and served by Torch.
- Startup export scales with the ladders: ≈2 s per shape, ≈35–45 s for the default 15 on Qwen3-0.6B. CUDA's own default ladders (≈42 prefill buckets, a memory-tiered decode cap) would take minutes, which is why MPS gets platform-sized defaults behind the same flags; `--cuda-graph-max-bs-{decode,prefill}` bound it.
- Radix-prefix prefills (a single request whose prompt hits a cached prefix) are served by the paged prefill kernel, which is the weaker of the two custom kernels at prefill shapes. Proposed per review: decline these batches to eager until the kernel-tuning PR, which removes that kernel from this PR.
- The exported logits path is hidden states times the LM head, narrowed to `config.vocab_size` the way `LogitsProcessor` narrows it; models needing more (logit scale, softcapping) are rejected rather than supported. Returning hidden states and running the real `LogitsProcessor` is the follow-up that lifts this and enables prompt logprobs.
- RoPE lowers as decomposed elementwise ops; pattern-matching it to `mx.fast.rope` is the known remainder of the prefill gap (~89 % of `mlx_lm`).
- MLA, sliding-window, SSM, and MoE architectures are rejected at startup by design and serve eagerly.
- Weight replacement invalidates executors through a storage-identity check (unit-tested); an end-to-end weight-update test through the update RPCs is still missing.

## Rollout plan

Tracked as checkboxes in the RFC's [Roadmap](https://github.com/sgl-project/sglang/issues/32321#user-content-roadmap).

1. **This PR** — the region as a bounded compute provider: additive, off by default, eager fallback for everything it does not serve.
2. **Native MLX lane removal** (2–3 PRs, after parity sign-off here) — `MlxTpModelWorker`, `MlxModelRunner`, its KV cache and scheduler mixin, `SGLANG_USE_MLX`; the ownership change the RFC argues for.
3. **Independent PRs, any order** — eager-path Metal ops (flag-off speedup), torch-native LoRA on MPS, platform device lifecycle; none is a dependency of the region.
4. **Kernel tuning + benchmarking PR** — paged prefill kernel, the block-table decode kernel and AOT Metal build from #32451 behind the same `mlx_radix_attention` façade, `index_copy_` fallback for the KV commit.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/srt/hardware_backend/mlx/fx_backend.py [hunk 1]
```diff
@@ -0,0 +1,1094 @@
...
+        for target in ("chunk", "reshape", "view", "transpose", "contiguous"):
+            registry.register_method(target, target)
+        return registry
+
+    @classmethod
+    def standard_export_decoder(cls) -> MlxFxLoweringRegistry:
+        """Return canonical ATen lowerings produced by ``torch.export``."""
+        registry = cls()
+        for target, lowering in (
+            (torch.ops.aten._assert_tensor_metadata.default, "assert_metadata"),
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_backend.py]: It would be better to store the mappings in a config file, so we wouldn’t need to modify the code whenever we want to update them.

File: python/sglang/srt/hardware_backend/mlx/fx_backend.py [hunk 2]
```diff
@@ -0,0 +1,1094 @@
...
+    lowering: str,
+    args: tuple[Any, ...],
+    kwargs: dict[str, Any],
+) -> Any:
+    import mlx.core as mx
+
+    if lowering == "assert_metadata":
+        return None
+    if lowering == "alias":
+        return args[0]
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/fx_backend.py]: This is really hard-coded. Is there a cleaner way to do this?

File: python/sglang/srt/hardware_backend/mlx/region_runner.py [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+# and the pools carry a +1 row at index 0 for it. Any other index -- including
+# ``token_to_kv_pool.size``, which is in bounds because of that extra row --
+# is allocatable, so pad rows would overwrite a live request's K/V with no
+# error once the pool neared exhaustion.
+_PAD_SINK_SLOT = 0
+
+# Decode shapes exported at startup under SGLANG_MLX_REGION_STARTUP_EXPORT.
+# Mirrors the spirit of the CUDA-graph capture ladder; sizes off the ladder
+# still export lazily at their first batch.
+_STARTUP_DECODE_BATCH_SIZES = (1, 2, 4, 8, 16)
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/region_runner.py]: Please check `server_args.py`, there should be parameters to control the cudagraph captured batch sizes.

Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/region_runner.py]: Done. the runner now reads `cuda_graph_config` like the other platform runners, so `--cuda-graph-bs-{decode,prefill}` / `--cuda-graph-max-bs-{decode,prefill}` work on MPS.

File: python/sglang/srt/hardware_backend/mlx/fx_backend.py (hunk 1 above)
Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_backend.py]: good idea, I will do it (Also thinking if this can be done in separate PR).

Comment (noob-se7en) [reply python/sglang/srt/hardware_backend/mlx/fx_backend.py]: Thought about this more, went with a decorator registry instead of a config file, and split it into its own PR as suggested: #37354. 
Each lowering function now declares its own targets via `@_lowering(name, aten=...)`, and both registries are derived from those declarations, so adding an op is one self-contained function with no table to keep in sync. 

Decided against a config file since every new op needs its MLX implementation written anyway: the file would only hold aliases and a typo'd op name in config fails silently at serve time, while a bad registration fails at import.
